File: app/src/main/java/dk/eightyplus/annoto/component/Picture.java
package dk.eightyplus.annoto.component;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.RectF;
import dk.eightyplus.annoto.R;
import dk.eightyplus.annoto.utilities.FileId;
import dk.eightyplus.annoto.utilities.Storage;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

/**
 * Picture component in the Composite pattern
 */
@SuppressWarnings("unused")
public class Picture extends Component {

  private transient Bitmap bitmap;
  private final String filename;
  private final transient Context context;

  private Picture(final Context context, final String filename) {
    this.context = context;
    this.filename = filename;
  }

  public Picture(final Context context, final Bitmap bitmap, final String filename) {
    this(context, filename);
    if (bitmap == null) {
      throw new NullPointerException(context.getString(R.string.log_error_bitmap_null));
    }
    this.bitmap = bitmap;
  }

  @Override
  public void onDraw(Canvas canvas, Paint paint) {
    if (isVisible()) {
      canvas.drawBitmap(bitmap, getMatrix(), paint);
    }
  }

  private Matrix getMatrix() {
    float[] values = new float[Matrix.MPERSP_2 + 1];
    values[Matrix.MSCALE_X] = values[Matrix.MSCALE_Y] = getScale();
    values[Matrix.MTRANS_X] = x;
    values[Matrix.MTRANS_Y] = y;
    values[Matrix.MPERSP_2] = 1;
    Matrix matrix = new Matrix();
    matrix.reset();
    matrix.setValues(values);
    return matrix;
  }

  @Override
  public float centerDist(float x, float y) {
    return calculateCenterDistance(x, y, getBounds());
  }

  @Override
  public RectF getBounds() {
    return new RectF(x, y, x + bitmap.getWidth() * scale, y + bitmap.getHeight() * scale);
  }

  @Override
  public boolean delete() {
    super.delete();
    return Storage.getStorage(context).deleteFile(filename);
  }

  @Override
  public ComponentType getType() {
    return ComponentType.PictureType;
  }

  public static Picture fromJson(final Context context, JSONObject object) throws JSONException {
    Picture picture = new Picture(context, object.getString(FileId.FILE_NAME));
    picture.fromJsonPrimary(object);
    return picture;
  }

  /**
   * Initialises picture when loaded from storage
   * @throws IOException
   */
  public Picture initialise() throws IOException {
    bitmap = Storage.getStorage(context).loadFromFile(filename);
    if (bitmap == null) {
      throw new IOException(context.getString(R.string.log_error_file_missing));
    }
    return this;
  }

  @Override
  public JSONObject toJson() throws JSONException {
    JSONObject object = super.toJson();
    object.put(FileId.FILE_NAME, filename);
    return object;
  }
}

File: app/src/main/java/dk/eightyplus/annoto/component/Picture.kt
package dk.eightyplus.annoto.component

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.RectF
import dk.eightyplus.annoto.R
import dk.eightyplus.annoto.utilities.FileId
import dk.eightyplus.annoto.utilities.Storage
import org.json.JSONException
import org.json.JSONObject

import java.io.IOException

/**
 * Picture component in the Composite pattern
 */
class Picture private constructor(@field:Transient private val context: Context, private val filename: String) : Component() {

    @Transient
    private var bitmap: Bitmap? = null

    private val matrix: Matrix
        get() {
            val values = FloatArray(Matrix.MPERSP_2 + 1).apply {
                this[Matrix.MSCALE_Y] = scale
                this[Matrix.MSCALE_X] = scale
                this[Matrix.MTRANS_X] = x
                this[Matrix.MTRANS_Y] = y
                this[Matrix.MPERSP_2] = 1f
            }
            return Matrix().apply {
                reset()
                setValues(values)
            }
        }

    override val bounds: RectF
        get() = RectF(x, y, x + (bitmap?.width ?: 0) * scale, y + (bitmap?.height ?: 0) * scale)

    override val type: ComponentType
        get() = ComponentType.PictureType

    constructor(context: Context, bitmap: Bitmap, filename: String) : this(context, filename) {
        this.bitmap = bitmap
    }

    override fun onDraw(canvas: Canvas, paint: Paint) {
        if (isVisible) {
            bitmap?.let {
                canvas.drawBitmap(it, matrix, paint)
            }
        }
    }

    override fun centerDist(x: Float, y: Float): Float {
        return calculateCenterDistance(x, y, bounds)
    }

    override fun delete(): Boolean {
        super.delete()
        return Storage.getStorage(context).deleteFile(filename)
    }

    /**
     * Initialises picture when loaded from storage
     * @throws IOException
     */
    @Throws(IOException::class)
    fun initialise(): Picture {
        bitmap = Storage.getStorage(context).loadFromFile(filename) ?: throw IOException(context.getString(R.string.log_error_file_missing))
        return this
    }

    @Throws(JSONException::class)
    override fun toJson(): JSONObject {
        return super.toJson().apply {
            put(FileId.FILE_NAME, filename)
        }
    }

    companion object {

        @Throws(JSONException::class)
        fun fromJson(context: Context, jsonObject: JSONObject): Picture {
            return Picture(context, jsonObject.getString(FileId.FILE_NAME)).apply {
                fromJsonPrimary(jsonObject)
            }
        }
    }
}
